dashboard/modules/serve/serve_agent.py

```diff
@@ -144,21 +165,42 @@ async def put_all_deployments(self, req: Request) -> Response:
 
         return Response()
 
-    def get_serve_client(self):
-        """Gets the ServeControllerClient to the this cluster's Serve app.
+    async def get_serve_controller(self):
+        """Gets the ServeController to the this cluster's Serve app.
 
         return: If Serve is running on this Ray cluster, returns a client to
             the Serve controller. If Serve is not running, returns None.
         """
+        async with self._controller_lock:
+            if self._controller is not None:
+                try:
+                    await self._controller.check_alive.remote()
+                    return self._controller
+                except ray.exceptions.RayActorError:
+                    logger.info("Controller is dead")
+                self._controller = None
+
+            # Try to connect to serve even when we detect the actor is dead
+            # because the user might have started a new
+            # serve cluter.
+            from ray.serve._private.constants import (
+                SERVE_CONTROLLER_NAME,
+                SERVE_NAMESPACE,
+            )
 
-        from ray.serve.context import get_global_client
-        from ray.serve.exceptions import RayServeException
-
-        try:
-            return get_global_client(_health_check_controller=True)
-        except RayServeException:
-            logger.debug("There's no Serve app running on this Ray cluster.")
-            return None
+            try:
+                self._controller = ray.get_actor(
```

Comment:
This will only find the controller when it's running in [`detached` mode](https://github.com/ray-project/ray/blob/a4434fac7f63b559e2fc5d763a505160dbae3558/python/ray/serve/_private/api.py#L182-L185). @edoakes Is that alright?

@iycheng Is there a reason we're adding this logic to the Serve agent instead of offloading it to the [`get_global_client`](https://github.com/ray-project/ray/blob/e5a8b1dd55817305ac85237cd99d8b6ed23df294/python/ray/serve/context.py#L36) function?

Reply:
The main reason is that get_global_client is not async which potentially will blocks serve agent and serve agent will blocks dashboard agent, and dashboard agent will blocks healthz, so k8s will kill the raylet.

I think there should be a refactoring or improvements of serve api to make them async but I notice it requires a lot of thing to do and probably not fit for 2.0 in both time constraints and complexity adds.